bqskit/ir/lang/qasm2/visitor.py

```diff
@@ -650,9 +646,77 @@ def measure(self, tree: lark.Tree) -> None:
                 'measured to a single classical bit.',
             )
 
+        params: list[float] = []
+        qlist = tree.children[0]
+        location = CircuitLocation(self.convert_qubit_ids_to_indices(qlist))
+
+        # Parse gate object
+        gate_name = tree.data
+        if gate_name in self.gate_defs:
+            gate_def: GateDef | CustomGateDef = self.gate_defs[gate_name]
+        elif gate_name in self.custom_gate_defs:
+            gate_def = self.custom_gate_defs[gate_name]
+        else:
+            raise LangException('Unrecognized gate: %s.' % gate_name)
+
+        if len(params) != gate_def.num_params:
+            raise LangException(
+                'Expected %d params got %d params for gate %s.'
+                % (gate_def.num_params, len(params), gate_name),
+            )
+
+        if len(location) != gate_def.num_vars:
+            raise LangException(
+                'Gate acts on %d qubits, got %d qubit variables.'
+                % (gate_def.num_vars, len(location)),
+            )
+
+        # Build operation and add to circuit
+        self.op_list.append(gate_def.build_op(location, params))
+
     def reset(self, tree: lark.Tree) -> None:
-        """Reset statement node visitor."""
-        raise LangException('BQSKit currently does not support resets.')
+        """Reset node visitor."""
+        params: list[float] = []
+        qlist = tree.children[-1]
+        gate_name = tree.data
+        if gate_name in self.gate_defs:
```

Comment:
Same comment as above. This entire method can be simplified a lot. We don't need to access or use gate_defs. We just need to parse the line into either a `reset qreg`, `reset qubit`, or an invalid qasm string. In each case, its directly clear how to add ops to the `op_list`.